lib/backend/bdb_ro.c

```diff
@@ -0,0 +1,811 @@
+{
+    if (cur->page)
+	free(cur->page);
+    if (cur->ovpage)
+	free(cur->ovpage);
+    if (cur->keyov.kv)
+	free(cur->keyov.kv);
+    if (cur->valov.kv)
+	free(cur->valov.kv);
+    free(cur);
+}
+
+static int cur_next(struct bdb_cur *cur)
+{
+    if (cur->state < 0)
+	return -1;
+    if (cur->db->type == BDB_HASH)
+	return hash_next(cur);
+    if (cur->db->type == BDB_BTREE)
+	return btree_next(cur);
+    return -1;
+}
+
+static int cur_getval(struct bdb_cur *cur)
+{
+    if (cur->state < 0)
+	return -1;
+    if (cur->db->type == BDB_HASH)
+	return hash_getval(cur);
+    if (cur->db->type == BDB_BTREE)
+	return btree_getval(cur);
+    return -1;
+}
+
+static int cur_lookup(struct bdb_cur *cur, const unsigned char *key, unsigned int keyl)
+{
+    int r = -1;
+    if (cur->db->type == BDB_HASH)
+	r = hash_lookup(cur, key, keyl);
+    if (cur->db->type == BDB_BTREE)
+	r = btree_lookup(cur, key, keyl);
+    if (r != 0)
+	return r;
+    cur->islookup = 1;
+    while ((r = cur_next(cur)) == 0)
+	if (keyl == cur->key.len && !memcmp(key, cur->key.kv, keyl))
+	    break;
+    cur->islookup = 0;
+    if (r == 0)
+	r = cur_getval(cur);
+    return r;
+}
+
+static int cur_lookup_ge(struct bdb_cur *cur, const unsigned char *key, unsigned int keyl)
+{
+    int r = -1;
+    if (cur->db->type == BDB_BTREE)
+	r = btree_lookup(cur, key, keyl);
+    if (r != 0)
+	return r;
+    cur->islookup = 1;
+    while ((r = cur_next(cur)) == 0) {
+	unsigned int ekeyl = cur->key.len;
+	int cmp = memcmp(cur->key.kv, key, keyl < ekeyl ? keyl : ekeyl);
+	if (cmp > 0 || (cmp == 0 && ekeyl >= keyl))
+	    break;
+    }
+    cur->islookup = 0;
+    if (r == 0)
+	r = cur_getval(cur);
+    else if (r == 1)
+	r = cur_next(cur);
+    return r;
+}
+
+/****** glue code ******/
+
+static unsigned int getui32(unsigned char *x, int swapped)
+{
+    union _dbswap bs;
+    memcpy(bs.uc, x, 4);
+    if (swapped)
+      _DBSWAP(bs);
+    return bs.ui;
+}
+
+static void setui32(unsigned char *x, uint32_t v, int swapped)
+{
+    union _dbswap bs;
+    bs.ui = v;
+    if (swapped)
+      _DBSWAP(bs);
+    memcpy(x, bs.uc, 4);
+}
+
+static void log_error(dbiIndex dbi)
+{
+    rpmlog(RPMLOG_ERR, "bdb_ro error reading %s database\n", dbi->dbi_file);
+}
+
+static int bdbro_Open(rpmdb rdb, rpmDbiTagVal rpmtag, dbiIndex * dbip, int flags)
+{
+    const char *dbhome = rpmdbHome(rdb);
+    dbiIndex dbi = NULL;
+    char *path;
+
+    if (dbip)
+	*dbip = NULL;
+    if ((rdb->db_mode & O_ACCMODE) != O_RDONLY)
+	return EPERM;
+    if ((dbi = dbiNew(rdb, rpmtag)) == NULL)
+	return 1;
+    path = rstrscat(NULL, dbhome, "/", dbi->dbi_file, NULL);
+    rpmlog(RPMLOG_DEBUG, "opening  db index       %s\n", path);
+    dbi->dbi_db = bdb_open(path);
+    if (!dbi->dbi_db) {
+	rpmlog(RPMLOG_ERR, "could not open %s: %s\n", path, strerror(errno));
+	free(path);
+	dbiFree(dbi);
+	return 1;
+    }
+    free(path);
+    dbi->dbi_flags |= DBI_RDONLY;
+    if (dbip)
+	*dbip = dbi;
```

Comment:
This would leak the dbi allocation if somebody actually called it with dbip as NULL. I don't think anything does, but the code checks for that case so it'll likely trip up false positives from static analyzers if nothing else.

Here too, as allocations are involved, I'd much prefer single point of cleanup + exit idiom that rpm generally uses. Not a showstopper though.